### DISCUSSION TITLE
求教几个疑问

### DISCUSSION THREAD

**Discussion by stopandjoin** [Tencent/tquic#195] [Q&A (问题和回答)] [answered]
有几个疑问，请教一下大家：
1，由于不懂Rust语言，所以直接用的C API。 C API的稳定性和性能 对比Rust而言， 是否有数据，或是有经过考验？
2，stream的优先级，不同优先级的传输策略是怎样的。是按照stream优先级以一定的比例分配总带宽，还是让优先级高的stream发完后，才能发低优先级的stream？
3，连接迁移是tquic内部已实现好了，对上层透明的 ？ 还是需要调用特定的API？

**Comment by glime-ai** [comment]
1. 关于C API的稳定性和性能对比Rust的数据，目前提供的信息中没有直接的性能比较数据或稳定性报告。TQUIC是基于Rust语言开发的，因此其Rust API可能会更加原生和优化。然而，提供C API是为了确保更广泛的兼容性和可用性，尽管可能在性能上有所差异，但TQUIC项目团队应该已经确保了其足够的稳定性和合理的性能。如果性能和稳定性对您的项目至关重要，建议进行具体的性能测试来评估C API与Rust API在您的使用场景中的表现。

2. 关于stream的优先级和传输策略，文档中没有提供具体的实现细节。一般来说，QUIC协议允许通过流优先级来指导数据的传输，但具体的策略（如是否按比例分配带宽或优先级高的流先发送）可能取决于QUIC库的实现。由于TQUIC是一个高性能的QUIC协议库，它可能实现了一种高效的流优先级处理机制，但具体细节需要查阅TQUIC的官方文档或直接咨询开发者。

3. 对于连接迁移的问题，QUIC协议支持连接迁移特性，允许连接在网络变化时保持活跃状态。TQUIC作为一个QUIC协议库，理论上应该支持这一特性。根据提供的信息，TQUIC的设计目标包括高性能和高质量，这意味着连接迁移特性可能已经被内部实现。然而，是否需要调用特定API来启用或管理连接迁移，或者这一过程对上层完全透明，需要进一步查阅TQUIC的文档或询问其开发者以获得确切信息。

总的来说，对于您的问题，建议直接查阅[TQUIC的官方文档](https://tquic.net/zh/docs/intro)或在其GitHub项目页面提出issue来获取最准确的答案。

**Comment by stopandjoin** [comment]
求教 @iyangsj   ^_^

**Reply by iyangsj** [reply, marked as answer]
1. 内部使用的也是C API

2. Stream优先级由 **紧急度**和 **增量标记** 构成，与HTTP3 优先级机制相同
```
int quic_stream_new(struct quic_conn_t *conn,
                    uint64_t stream_id,
                    uint8_t urgency,
                    bool incremental);
```

简述如下：具有数据待发送的流，首先按优先级分为不同队列；相同优先级的流再按 non-incremental/incremental分为两类
<img width="557" alt="image" src="https://github.com/Tencent/tquic/assets/780807/c2519c9a-3b79-4460-8763-d1a811564fdb">

- 发送报文时，优先处理最高优先级流的缓冲数据（即urgency 最小的流），处理完成后才依次处理更低优先级的流

- 相同优先级的流可能存在多个，按如下方式处理
  - 优先处理non-incremental类型的流缓冲数据，按流ID从小到大顺序进行调度（即流ID更小的缓冲数据发送完，才处理下一个）
  - 然后处理incremental类型的流缓冲数据，按轮转方式依次调度发送（即交错发送不同流的数据）

例：JS会阻塞渲染，使用高优先级及non-incremental标记；而对于渐进式 jpg图片，可以使用普通优先级及incremental标记交错传输

https://github.com/Tencent/tquic/blob/develop/src/connection/stream.rs#L3197-L3207

3. 开源版本暂不支持连接迁移，后续版本会支持，客户端需要调用migrate()接口将连接迁移到期望的四元组

**Reply by stopandjoin** [reply]
好的，很清楚啦，谢谢。

**Reply by stopandjoin** [reply]
@iyangsj 关于连接迁移，还有个几个问题确认一下： 比如手机移动位置后，网络从一个基站切换到另一个基站， 这种情况下：
1，tquic内部会自动重连吗，还是需要业务层去做重连逻辑？ 
2，这种重连概念（只需知道target ip和port）和连接迁移（需要四元组） 有区别吧，如果客户端能及时拿到自己本地新的ip 和port，调用将来的migrate()接口，就能无感迁移，不用重连？

**Reply by iyangsj** [reply]
1.需要业务层重新创建连接，并重试失败的请求（在未启用连接迁移或多路径模式下）
2.在启用连接迁移情况下，客户端调用migrate接口(提供新的本地新的ip 和port)，会执行连接迁移，无需重连

**Comment by cmsly** [comment]
@iyangsj 下面这个接口有实现时间计划吗？
/// Migrates the connection to the specified path.
    #[doc(hidden)]
    pub fn migrate_path(&mut self, local_addr: SocketAddr, remote_addr: SocketAddr) -> Result<()> {
        // TODO: support migration
        Err(Error::InternalError)
    }
